File: pyscripts/call_limit.py
#!/usr/bin/python
##Polling file
import time
import freeswitch
import re
import random

rootdir = freeswitch.getGlobalVariable("base_dir");
api = freeswitch.API();
file_path = rootdir+"/scripts/pyscripts/sipnumber/number"

def file_list(filename):
    num_list = []
    file = open(filename,"r")
    for line in file.readlines():
        line = line.strip('\n')
        num_list.append(line)
    file.close()
    return num_list

def check_gateway_limits(key, limits, upgw):
    count = 0
    today = time.strftime("%Y%m%d", time.localtime())
    if not api.executeString("db exists/call_limit_last_count_of_day/data_time"):
        api.executeString("db insert/call_limit_last_count_of_day/data_time/"+str(today))
    else:
        last_day = api.executeString("db select/call_limit_last_count_of_day/data_time")
    if not api.executeString("db select/call_limit_value_count_of_day/"+str(key)):
        api.executeString("db insert/call_limit_value_count_of_day/"+str(key)+"/0")
    else:
        count = api.executeString("db select/call_limit_value_count_of_day/"+str(key))
    if today != last_day:
        api.executeString("db delete/call_limit_value_count_of_day/data_time")
        for k in range(len(upgw)):
            api.executeString("db delete/call_limit_value_count_of_day/"+str(upgw[k]))
    if int(count) < int(limits):
        count = int(count) + 1
        api.executeString("db insert/call_limit_value_count_of_day/"+str(key)+"/"+str(count))
        api.executeString("db insert/call_limit_last_count_of_day/data_time/"+str(today))
        return True
    else:
        return False

def db_insert(current_pos):
     api.executeString("db insert/gateway_route/current_pos/"+str(current_pos))

def db_select():
    return api.executeString("db select/gateway_route/current_pos")

def db_exists():
    if not api.executeString("db select/gateway_route/current_pos"):
        api.executeString("db insert/gateway_route/current_pos/0")

def handler(session, args):
    pass

def fsapi(session, stream, env, args):
    count, fullgw, current_pos, new_pos, idx = 0, 0, 0, 0, 0
    upgw = []
    if args:
        string = re.split('\s+', args)
    else:
        freeswitch.consoleLog('info', 'must argv1 profile %s\n' % args)
    argv1 = string[0]
    upgw_list = file_list(file_path)
    for gw in range(len(upgw_list)):
        if re.search(r"State\s+REGED\s+",api.executeString("sofia status gateway "+upgw_list[gw])):
            upgw.append(upgw_list[gw])
            count = count + 1
    db_exists()
    current_pos = db_select()
    if int(current_pos) >= count:
        current_pos = 0
    new_pos = int(current_pos) + 1
    idx = int(current_pos)
    freeswitch.consoleLog('info', 'the idx is %s\n' % idx)
    freeswitch.consoleLog('info', 'the upgw is %s\n' % upgw[idx])
    db_insert(new_pos)
    info = api.executeString("show channels")
    while 0 < 1:
        if re.search("sofia/gateway/"+upgw[idx],info):
            idx = idx + 1
            if int(current_pos) >= count:
                idx = 0
            if idx >= count:
                break
        else:
            if check_gateway_limits(upgw[idx],argv1,upgw):
                stream.write(str(upgw[idx]))
                freeswitch.consoleLog("info", "Discovery of available gateway: %s\n" % upgw[idx])
                break
            else:
                freeswitch.consoleLog("err", "unavailable gateway\n")
                break

File: freeswitch_lua/轮询/call_limit.lua
---顺序轮序,限制呼叫次数
local rootdir = freeswitch.getGlobalVariable("base_dir");
local api =  freeswitch.API();


function string_spilt(str,sep)
	if str == nil or sep == nil then 
		return {};
	end
	local s = 1;
	local arr = {};
	local c = 1;
	while true do
		local s1 = string.find(str,sep,s);		
		if s1 == nil then
			arr[c] = string.sub(str,s,-1);
			break;
		end
		if s == s1 then
			arr[c] = "";
		else
			arr[c] = string.sub(str,s,s1-1); 
		end		
		s = s1 + 1;
		c = c+1;
	end
	return arr;
end


function check_gateway_limits(key,limits)
	local last = nil
	if api:executeString("db exists/call_limit_last_count_of_day/data_time") == "false" then
	    	local last_day = os.date("%Y%m%d")
	    	api:executeString("db insert/call_limit_last_count_of_day/data_time/"..last_day)
	else
        	last_day = api:executeString("db select/call_limit_last_count_of_day/data_time")
    end
	local count = 0
	if api:executeString("db exists/call_limit_value_count_of_day/"..key) == "false" then
		api:executeString("db insert/call_limit_value_count_of_day/"..key.."/0")
	else
        	count = tonumber(api:executeString("db select/call_limit_value_count_of_day/"..key)) 	
    	end
	local current_time = os.date("%Y%m%d")
	if current_time ~= last_day then
		count = 0
		api:executeString("db delete/call_limit_last_count_of_day/data_time")
		local keys = string_spilt(api:executeString("db list/call_limit_value_count_of_day/"),",")
		for k,v in ipairs(keys) do
            		api:executeString("db delete/call_limit_value_count_of_day/"..v)		
		end	
		api:executeString("db insert/call_limit_last_count_of_day/data_time/"..current_time)
	end
	freeswitch.consoleLog("info","Limit the number of calls per day,'"..key.."',limits:"..limits..",count:"..count..".\n")
	if count < limits then
		count = count + 1 
		api:executeString("db insert/call_limit_value_count_of_day/"..key.."/"..count)
		return true
	else
	    	return false
   	end
	
end

file_path=rootdir.."/scripts/route/number.txt"
local gw = {}
local index = 1
local max = 0
local profile = nil
profile = argv[1]
if argv[1] == nil or string.match(profile,"qidian") == nil then
        freeswitch.consoleLog("ERR","must argv[1] profile!\n");
        return
end
if argv[2] == nil then
	freeswitch.consoleLog("ERR","must argv[2] limits!\n")
	return 
end 
local limits = tonumber(argv[2])
local myfile = io.open(file_path,"r")
for i in myfile:lines() do
	gw[index] = i
	index = index+1
end
myfile:close()
if api:executeString("db exists/gateway_route/current_pos") == "false" then
	api:executeString("db insert/gateway_route/current_pos/0")
end

count = 0
upgw = {}
gwstatusstr = api:executeString("sofia status gateway")
for k,v in ipairs(gw) do
	if string.match(api:executeString("sofia status gateway "..v),"Status%s+UP%s+") then
		upgw[count] = v
		count = count+1
	end	
end
current_pos = tonumber(api:executeString("db select/gateway_route/current_pos")) 
if current_pos>=count then
	current_pos = 0
end

new_pos = current_pos + 1
api:executeString("db insert/gateway_route/current_pos/"..new_pos)
info = api:executeString("show channels")
idx = current_pos
while true do	
	st,_ = string.find(info,"sofia/gateway/"..upgw[idx])
	if st==nil then
		if check_gateway_limits(upgw[idx],limits) then
			stream:write(upgw[idx])
			break
		end
	end
	if idx == current_pos-1 then
		freeswitch.consoleLog("WARNING","No available gateways have been found.\n");
		break
	end
	idx = idx+1
	if current_pos>=count then
		idx = 0
	end
	if idx >= count then
		break
	end
end


